### PR TITLE
PERF-#4804: Preserve lengths/widths caches in `broadcast_apply_full_axis`

### LINKED ISSUE #4804: PERF: `broadcast_apply_full_axis` can be implemented to compute caches in some cases

### PR DESCRIPTION
**Pull request by anmyachev**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4804 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 1]
```diff
@@ -3348,6 +3348,25 @@ def broadcast_apply_full_axis(
                         kw["column_widths"] = self._column_widths_cache
                     elif len(new_columns) == 1 and new_partitions.shape[1] == 1:
                         kw["column_widths"] = [1]
+        else:
+            if (
+                kw["row_lengths"] is None
```
Comment (dchigarev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:3353]:
```suggestion
                axis == 0 and
                kw["row_lengths"] is None
```

It seems that we can preserve lengths only for one axis (for the one, that we're applying a full-axis function). Consider this case, where kernel modifies the shapes of partitions without modifying the total length:
```python
from modin.test.storage_formats.pandas.test_internals import construct_modin_df_by_scheme
import pandas

md_df = construct_modin_df_by_scheme(pandas.DataFrame({"a": [1, 2, 3, 4], "b": [3, 4, 5, 6], "c": [6, 7, 8, 9], "d": [0, 1, 2, 3]}), {"row_lengths": [2, 2], "column_widths": [2, 2]})._query_compiler._modin_frame

def func(df):
    if df.iloc[0, 0] == 1:
        return pandas.DataFrame({"a": [1, 2, 3], "b": [2, 3, 4], "c": [4, 2, 1], "d": [3, 2, 1]})
    else:
        return pandas.DataFrame({"a": [3], "b": [4], "c": [1], "d": [1]})

res = md_df.apply_full_axis(func=func, axis=1, new_index=[0, 1, 2, 3], new_columns=["a", "b", "c", "d"], keep_partitioning=True)
print(res._row_lengths_cache) # [2, 2]
print(res._column_widths_cache) # [2, 2]

actual_row_lengths = [part.length() for part in res._partitions[:, 0]]
print(actual_row_lengths) # [3, 1]
```

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 2]
```diff
@@ -4156,21 +4155,7 @@ def compute_groupby(df, drop=False, partition_idx=0):
...
             apply_indices = None
 
-        if (
-            # For now handling only simple cases, where 'by' columns are described by a single query compiler
-            agg_kwargs.get("as_index", True)
-            and len(not_broadcastable_by) == 0
-            and len(broadcastable_by) == 1
-            and broadcastable_by[0].has_materialized_dtypes
-        ):
-            new_index = ModinIndex(
```
Comment (anmyachev) [inline modin/core/storage_formats/pandas/query_compiler.py:4166]: @dchigarev  How is it supposed to take the length from this object in `broadcast_apply_full_axis` function if it refers to the old modin frame?

Comment (anmyachev) [reply modin/core/storage_formats/pandas/query_compiler.py]: @dchigarev ready for review

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 3]
```diff
@@ -3272,6 +3272,27 @@ def broadcast_apply_full_axis(
...
                         kw["column_widths"] = [1]
+        else:
+            if (
+                axis == 0
+                and kw["row_lengths"] is None
+                and self._row_lengths_cache is not None
+                and ModinIndex.is_materialized_index(new_index)
+                and len(new_index) == sum(self._row_lengths_cache)
+                # to avoid problems that may arise when filtering empty dataframes
+                and all(r != 0 for r in self._row_lengths_cache)
```
Comment (dchigarev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:3283]: what kind of problems do you mean here?

File: modin/test/storage_formats/pandas/test_internals.py [hunk 4]
```diff
@@ -1398,6 +1398,70 @@ def test_sort_values_cache():
...
+        func=func,
+        axis=1,
+        new_index=[0, 1, 2, 3],
+        new_columns=["a", "b", "c", "d"],
+        keep_partitioning=True,
+    )
+
+    actual_column_widths = [part.width() for part in res._partitions[0]]
+
+    assert res._column_widths_cache == actual_column_widths
```
Comment (dchigarev) [inline modin/test/storage_formats/pandas/test_internals.py:1430]: just to make sure that calling `.width()` doesn't propagate the cache anyhow to a parent dataframe
```suggestion
    col_widths_cache = res._column_widths_cache
    actual_column_widths = [part.width() for part in res._partitions[0]]

    assert col_widths_cache == actual_column_widths
```

File: modin/test/storage_formats/pandas/test_internals.py [hunk 5]
```diff
@@ -1398,6 +1398,70 @@ def test_sort_values_cache():
...
+        func=func,
+        axis=0,
+        new_index=[0, 1, 2, 3],
+        new_columns=["a", "b", "c"],
+        keep_partitioning=True,
+    )
+
+    actual_row_lengths = [part.length() for part in res._partitions[:, 0]]
+
+    assert res._row_lengths_cache == actual_row_lengths
```
Comment (anmyachev) [inline modin/test/storage_formats/pandas/test_internals.py:1461]:
```suggestion
    row_lengths_cache = res._row_lengths_cache
    actual_row_lengths = [part.length() for part in res._partitions[:, 0]]

    assert row_lengths_cache == actual_row_lengths
```

**Pull request merged**